### PR TITLE
Ocelots - Kelly T

### PR DESCRIPTION
**Pull request by kellytate**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -119,12 +119,14 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    # Added Assert
+    assert updated_data["watched"][-1]["title"] == MOVIE_TITLE_1
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:127]: Nice assertion for the title! 

Since we have the assertion above `assert len(updated_data["watched"]) == 1` that tells us there is only 1 item in the `watched` list, we could also use `updated_data["watched"][0]` to access the movie we want to check.

In this case where the code is checking individual keys of a dictionary, I recommend including assertions for all of the relevant keys.  
```python
    watched_movie = updated_data["watched"][0]
    assert watched_movie["title”] == MOVIE_TITLE_1
    assert watched_movie["rating"] == RATING_1
    assert watched_movie["genre"] == GENRE_1
```

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -143,12 +145,14 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    # Added Assert
+    assert updated_data["watched"][-1]["title"] == MOVIE_TITLE_1
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:153]: The README isn't explicit about what order we should add movies to the list `watched`. If we want our tests to be independent from a specific implementation that adds movies to a particular location in the `watched` list, we could use the `in` keyword to check if the variable `movie_to_watch` is in the `watched` list instead of looking at a specific index. By checking for the whole dictionary, there's a side benefit that we don't need to check each key independently:
```python
assert movie_to_watch in updated_data["watched"]

# Another option:
assert HORROR_1 in updated_data["watched"]
```

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -55,12 +55,14 @@ def test_friends_unique_movies_not_duplicated():
...
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
-    # **************************************************************************************************
+    # ************************************************
+    
+    assert len(set(movie["title"] for movie in friends_unique_movies)) == len(friends_unique_movies)
```
Comment (kelsey-steven-ada) [inline tests/test_wave_03.py:63]: Nice use of a list comprehension to create the set here!

This assertion will create a set from the movie titles in `friends_unique_movies` and check that the length of that set is the same as the length of `friends_unique_movies`. This will ensure we don't have duplicates, but it doesn't confirm that `friends_unique_movies` holds only the movies we expect. I would suggest also checking if the specific movies we intend to be in `friends_unique_movies` are present. 

What could those assertions look like?

Comment (kelsey-steven-ada) [reply tests/test_wave_03.py:63]: The PEP 8 style guide recommends limiting lines to a max of 79 characters to make code easier to read, especially with multiple editor windows open: https://peps.python.org/pep-0008/#maximum-line-length. Another option here could be to create the set on a separate line, then use it in the `assert`:

```python
title_set = set(movie["title"] for movie in friends_unique_movies)
assert len(title_set) == len(friends_unique_movies)
```

File: tests/test_wave_05.py [hunk 4]
```diff
@@ -53,12 +53,18 @@ def test_new_genre_rec_from_empty_friends():
...
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    # raise Exception("Test needs to be completed.")
     # *********************************************************************
     # ****** Complete the Act and Assert Portions of theis tests **********
     # *********************************************************************
 
-@pytest.mark.skip()
+    # Assert
+    assert len(recommendations) == 0
```
Comment (kelsey-steven-ada) [inline tests/test_wave_05.py:65]: Great act & assert steps!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,139 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    
+    if title == None or genre == None or rating == None:
+        return None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:6]: Nice handling to return `None` for missing data! In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.
```python
if title is None or genre is None or rating is None :
        return None
```
If you’d like more info on `is` vs `==` for `None`, here’s a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html


Something else to consider: here we're explicitly checking if the values are None, but what if we got an empty string for `title`? If we wanted to check for empty strings at the same time as we check for None, we could rely on the truthy/falsy values of the parameters and write something like: 
```python
if not title or not genre or not rating:
    return None
```

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,139 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    
+    if title == None or genre == None or rating == None:
+        return None
+    else:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:7]: Since it doesn't change the result of our function, I recommend removing the `else` so we can outdent the main flow of the function.

If an `else` clause doesn't change the flow of our code, then we have a statement that needs to be compiled and evaluated, but that has no benefit. Additionally, they might cause the main flow of our code to be indented, when we generally want that to be prominent to readers. I generally recommend always questioning if we need an `else` clause, and only adding one if it's required.

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,139 @@
...
+
+def watch_movie(user_data, movie):
+
+    for movie_entry in user_data["watchlist"]:
+        
+        if movie_entry["title"] == movie:
+            user_data["watched"].append(movie_entry)
+            user_data["watchlist"].remove(movie_entry)
+    
+    return user_data
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:31]: This function works, but as an aside it contains an "anti-pattern" that is important to be aware of.  An element of the list that is being iterated over is being removed inside the for loop.  This won't cause problems in this particular case, but consider this slightly different implementation:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]

        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)

    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first is the movie that we are moving to `user_data["watched"]`, the second iteration through the loop will throw an index error, because it will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

As a coding pattern, this is one that I recommend avoiding because it has so much potential for unforeseen errors. One fix is to return immediately after making the swap (ie, add `return user_data` inside the if-block so that the loop ends as soon as the swap is made).  Another fix is to use the loop to find the item, but not make the swap until after the loop ends.

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,139 @@
...
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    
+    if len(user_data["watched"]) != 0:
+        sum = 0
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:39]: Something to note: the names `sum`, `max`, `min` (and several others) are used by a built in functions of python. We want to avoid using the names of built in functions for our variables because doing so replaces the value of the variable which the built in function was using - meaning we can lose access those built in functions because we no longer have a reference to them.

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,139 @@
...
+    
+        for movie in user_data["watched"]:
+            sum += movie["rating"]
+        return sum / len(user_data["watched"])
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:43]: Nice approach! Another way we could do this is using list comprehensions and the `sum` function:

```python
# The list comprehension inside the `sum` call pulls the value of the 
# `rating` key out of each item in the list `user_data['watched']`
sum = sum(movie["rating"] for movie in user_data["watched"])
return sum / len(user_data["watched"])
```

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,139 @@
...
+    
+    else:
+        return 0
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:46]: This is a little bit personal taste, but I often prefer to reverse the validation check and use it like a guard clause at the top of the function. This can let us take the main logic of our function out of an if-statement so it is prominent and clear rather than indented. For me, this makes code easier to read, but you may feel differently or work on a team that has guidelines for how they structure validation checks.

We can also use the truthy/falsy value of `user_data["watched"]` to confirm it is not empty and does not point to `None`:
```python
if not user_data["watched"]:
    return 0

sum = 0
for movie in user_data["watched"]:
    sum += movie["rating"]

return sum / len(user_data["watched"])
```

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,139 @@
...
+    if len(user_data["watched"]) != 0:
+        genres = {}
 
+        for movie_entry in user_data["watched"]:
+            if movie_entry["genre"] in genres:
+                genres[movie_entry["genre"]] += 1
+            else:
+                genres[movie_entry["genre"]] = 1
+        
+        return max(genres, key=genres.get)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:59]: Great use of a frequency map and `max`!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,139 @@
...
+def get_most_watched_genre(user_data):
+    
+    if len(user_data["watched"]) != 0:
+        genres = {}
 
+        for movie_entry in user_data["watched"]:
+            if movie_entry["genre"] in genres:
+                genres[movie_entry["genre"]] += 1
+            else:
+                genres[movie_entry["genre"]] = 1
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:57]: There's a tradeoff of using a little more space for another variable, but to make the code a little easier to read and reduce how many times we need to access `movie_entry["genre"]`, I might suggest creating a variable before the `if` that holds the result of `movie_entry["genre"]` which we can use on the subsequent lines:
```python
for movie_entry in user_data["watched"]:
    movie_genre = movie_entry["genre"]
    if movie_genre in genres:
        genres[movie_genre] += 1
    else:
        genres[movie_genre] = 1
```

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,139 @@
...
+    
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+
+    unique_movies = []
+
+    if len(user_data["watched"]) != 0:
+        friends_watched_titles = set()
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:73]: Nice use of `set`!

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,139 @@
...
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    
+    friends_unique_movies = []
+
+    if len(user_data["watched"]) != 0 and len(user_data["friends"]) != 0:
+        for friend in user_data["friends"]:
+            for movie_entry in friend["watched"]:
+                if movie_entry not in user_data["watched"] and movie_entry not in friends_unique_movies:
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:91]: The PEP 8 style guide recommends limiting lines to a max of 79 characters to make code easier to read, especially with multiple editor windows open: https://peps.python.org/pep-0008/#maximum-line-length. We would want to break it over a couple lines or create variables to hold the comparisons that we then use in an `if` statement:

```python
is_movie_in_my_list = movie_entry in user_data["watched"]
is_movie_in_unique_list = movie_entry in friends_unique_movies
if not is_movie_in_my_list and not is_movie_in_unique_list:
    friends_unique_movies.append(movie_entry)
```

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,139 @@
...
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    recommendations = []
+
+    if len(user_data["watched"]) != 0 and len(user_data["friends"]) != 0: 
+        possible_recs = get_friends_unique_watched(user_data)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:106]: Nice code reuse!

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,139 @@
...
+
+        recommendations = [movie for movie in possible_recs if movie["host"] in user_data["subscriptions"]]
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:108]: Great place for a list comprehension! I would suggest breaking this over a couple lines:
```python
recommendations = [movie for movie in possible_recs 
                  if movie["host"] in user_data["subscriptions"]]
```

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,139 @@
...
+    possible_recs = get_available_recs(user_data)
+
+    user_genres = set()
+
+    for movie in user_data["watched"]:
+        user_genres.add(movie["genre"])
+    
+    recommendations  = [movie for movie in possible_recs if movie["genre"] in user_genres]
+
+    return recommendations
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Though this solution passes the current test suite, it isn't doing what the problem statement asks. 

The test `test_new_genre_rec` uses `clean_wave_5_data()` as the input, which includes subscription data. If we removed "hulu" from the user's subscriptions on line 157 of `test_constants.py`, suddenly the test `test_new_genre_rec` starts failing. Our problem statement for `get_new_rec_by_genre` does not ask us to consider subscriptions though, so changing that subscription data should not change the result of `get_new_rec_by_genre`. 

What steps can you take to diagnose what is happening? 
How do we make sure that we are only considering movies where the `genre` of the movie is the same as the user's most watched genre?

**Review by kelsey-steven-ada — COMMENTED** [review]
🎉 Congrats on completing your first project at Ada! 🎉 I’ve added some suggestions & questions, let me know if there's anything I can clarify.

File: viewing_party/party.py (hunk 17 above)
Comment (kellytate) [reply viewing_party/party.py]: I misread the instructions (read 'genres' instead of 'genre') and made a flawed assumption that the get_new_rec_by_genre function should build on the previous function and require matching subscription. First troubleshooting step should have been to reread the problem statement carefully. Doing so revealed that I needed to check the possible_recs against the result of the get_most_watched_genre function.

Comment (kelsey-steven-ada) [reply viewing_party/party.py]: Nice investigation and updates!
